elastic/kibana
# [Security Solution] Siem signals -> alerts as data field and index aliases

## Summary

Addresses #100103 

The purpose of this work is to make the .siem-signals indices forwards-compatible and the alerts-as-data indices backwards-compatible. The field aliases introduced in this PR map the `signal` fields to their corresponding `kibana.alert` counterparts (if they exist). Some deprecated `signal` fields such as `signal.parent` will be removed in `kibana.alert` so they won't be available anymore. Forwards-compatibility on .siem-signals means we only have to maintain a single set of queries/visualizations in our UI and they will still work on old data. Backwards compatibility on alerts as data allows existing user rules, queries, and visualizations against .siem-signals to work seamlessly against the new alerts-as-data indices as well.

Testing:
Adding the aliases is done by visiting the Alerts page in the Security Solution with the proper permissions (the same way an admin visits the page initially to create the .siem-signals resources). Admins do not need any permissions beyond what is currently required (https://www.elastic.co/guide/en/security/7.x/detections-permissions-section.html).

Done:
- Create field aliases mapping signals field names to new alerts as data field names (so new visualizations work on old data)
- Create field aliases mapping alerts as data field names to signals field names (so existing visualizations continue to work on new data)
- Add alerts-as-data index aliases to existing .siem-signals indices (so queries on new indices find old data as well)
- Update existing siem signals templates to ensure all siem signals indices will have the correct mappings, even if the siem signals index rollsover after upgrade
- Replace legacy siem signal templates with new component template style
- Add RBAC required fields to .siem-signals indices as runtime fields or const keyword so RBAC works on old data
  - space_ids: const_keyword, consumer: const_keyword, producer: const_keyword, rule type: const_keyword

Follow up work:
- [ ] Once https://github.com/elastic/kibana/pull/105096 is merged, add the generated aliases into the alerts as data mappings
- [ ] Add .siem-signals index aliases to the new alerts-as-data indices (so existing queries and privileges will work on new data)
  - RuleDataClient should take an optional additional alias to add to bootstrapped index
  - When bootstrapping index, RuleDataClient needs to also create a template for the specific alias in order to specify the lifecycle policy and rollover_alias for the concrete indices backing the alias

### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/master/RISK_MATRIX.mdx) |


### For maintainers
